src/veh_interact.cpp

```diff
@@ -934,79 +928,180 @@ void veh_interact::do_mend()
+                    auto stack = units::legacy_volume_factor / item::find_type( pt.ammo_current() )->stack_size;
+                    right_print( w, y, 1, item::find_type( pt.ammo_current() )->color,
+                                 "%s  %5.1fL", item::nname( pt.ammo_current() ).c_str(),
+                                 round_up( to_liter( pt.ammo_remaining() * stack ), 1 ) );
+                }
+            };
+            opts.push_back( part_option { "TANK", &pt, enable ? enable( pt ) : false, details } );
+        }
     }
 
-    std::vector<vehicle_part *> ptanks;
-    auto opts = g->u.items_with( []( const item &e ) { return e.is_ammo(); } );
-    for( auto idx : parts_here ) {
-        if( std::any_of( opts.begin(), opts.end(), [&]( const item *e ) {
-            return veh->parts[ idx ].can_reload( e->typeId() );
-        } ) ) {
-            ptanks.push_back( &veh->parts[ idx ] );
+    for( auto &pt : veh->parts ) {
+        if( pt.is_battery() && !pt.is_broken() ) {
+            // always display total battery capacity and percentage charge
+            auto details = []( const vehicle_part &pt, WINDOW *w, int y ) {
+                int pct = ( double( pt.ammo_remaining() ) / pt.ammo_capacity() ) * 100;
+                right_print( w, y, 1, c_yellow, "%i    %3i%%", pt.ammo_capacity(), pct );
+            };
+           opts.push_back( part_option { "BATTERY", &pt, enable ? enable( pt ) : false, details } );
         }
-    };
+    }
 
-    if (ptanks.empty()) {
-        debugmsg("veh_interact::do_refill: Refillable parts list is empty.\n"
-                 "veh_interact::cant_do() should control this before.");
-        return;
+    for( auto &pt : veh->parts ) {
+        if( pt.is_reactor() && !pt.is_broken() ) {
+            // if reactor is fuelled then display contents in scaled plutonium units
+            auto details = []( const vehicle_part &pt, WINDOW *w, int y ) {
+                if( pt.ammo_remaining() ) {
+                    right_print( w, y, 1, item::find_type( pt.ammo_current() )->color,
+                                 "%s    %4i", item::nname( pt.ammo_current() ).c_str(), pt.ammo_remaining() );
+                }
+            };
+            opts.push_back( part_option { "REACTOR", &pt, enable ? enable( pt ) : false, details } );
+        }
+    }
+
+    for( auto &pt : veh->parts ) {
+        if( pt.is_turret() && !pt.is_broken() ) {
+            // if turret is loaded then display details about contained ammo
+            auto details = [&]( const vehicle_part &pt, WINDOW *w, int y ) {
+                if( pt.ammo_current() != "null" ) {
+                    right_print( w, y, 1, item::find_type( pt.ammo_current() )->color,
+                                 "%s   %5i", item::nname( pt.ammo_current() ).c_str(),
+                                 pt.ammo_remaining() );
+                }
+            };
+            opts.push_back( part_option { "TURRET", &pt, enable ? enable( pt ) : false, details } );
```

Comment:
Actually, the whole block is duplicated four times with minimal variations. It's worth refactoring/extracting, because the pattern is clearly seen:
```c++
for( auto &pt : veh->parts ) {
    if( pt.IS_PARTNAME() && !pt.is_broken() ) {
        // ... minimal variations ...
        opts.push_back( part_option { "PARTNAME" ... }
    }
}
```